### PR TITLE
part-1

### PR DESCRIPTION
**Pull request by d90375**
part-1

### CODE REVIEW

File: task/01-strings-tasks.js [hunk 1]
```diff
@@ -81,7 +78,7 @@ function  extractNameFromTemplate(value) {
  *   'cat'       => 'c'
  */
 function getFirstChar(value) {
-  throw new Error('Not implemented');
+  return value.slice(0, 1);
```
Comment (RamanRudski) [inline task/01-strings-tasks.js]: better get value by index like `value[0]`

File: task/01-strings-tasks.js [hunk 2]
```diff
@@ -199,10 +195,12 @@ function extractEmails(str) {
  *
  */
 function getRectangleString(width, height) {
-  throw new Error('Not implemented');
+  const topBorder = repeatString('─', width - 2);
```
Comment (RamanRudski) [inline task/01-strings-tasks.js]: please, rename it. for example `verticalBorder`

File: task/01-strings-tasks.js [hunk 3]
```diff
@@ -266,7 +264,9 @@ function isString(value) {
  *   'K♠' => 51
  */
 function getCardId(value) {
-  throw new Error('Not implemented');
+  const rowCount = 13; const suit = ['♣', '♦', '♥', '♠'];
```
Comment (RamanRudski) [inline task/01-strings-tasks.js]: You needn't it `const rowCount = 13`.

File: task/01-strings-tasks.js [hunk 4]
```diff
@@ -266,7 +264,9 @@ function isString(value) {
...
+  const n = ['A', '2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K'];
+  return suit.indexOf(value.slice(-1))*rowCount + n.indexOf(value.slice(0, -1));
```
Comment (RamanRudski) [inline task/01-strings-tasks.js]: `value` always has a length of 2. In this case, better to get the value by index. 
1. replace `slice()`
2. replace `rowCount` to `n.length`

File: task/02-numbers-tasks.js [hunk 5]
```diff
@@ -73,7 +70,7 @@ function getAverage(value1, value2) {
  *   (4, 2) (1, 6) => 5
  */
 function getDistanceBetweenPoints(x1, y1, x2, y2) {
-  throw new Error('Not implemented');
+  return Math.sqrt((Math.pow((x2 - x1), 2)) + (Math.pow((y2 - y1), 2)));
```
Comment (RamanRudski) [inline task/02-numbers-tasks.js]: You should use `Math.hypot` it is easier

File: task/02-numbers-tasks.js [hunk 6]
```diff
@@ -128,10 +124,9 @@ function getAngleBetweenVectors(x1, y1, x2, y2) {
  *     0     => 0
  */
 function getLastDigit(value) {
-  throw new Error('Not implemented');
+  return +value.toString().split('').reverse()[0];
```
Comment (RamanRudski) [inline task/02-numbers-tasks.js]: it is extra. You should think how use `%` in this case.

File: task/02-numbers-tasks.js [hunk 7]
```diff
@@ -203,7 +198,15 @@ function roundToPowerOfTen(num, pow) {
  *   17 => true
  */
 function isPrime(n) {
-  throw new Error('Not implemented');
+  if (n === 1) return false;
+  if (n === 2) return true;
+  if (n % 2 === 0) return false;
+  let bool = true;
+  for (let i = 3; i <= Math.floor(Math.sqrt(n)); i += 2) {
```
Comment (RamanRudski) [inline task/02-numbers-tasks.js]: remove first four string;
change for like `et i = 2; i < n; i += 1`

File: task/02-numbers-tasks.js [hunk 8]
```diff
@@ -203,7 +198,15 @@ function roundToPowerOfTen(num, pow) {
...
+    if (n % i === 0) bool = false;
```
Comment (RamanRudski) [inline task/02-numbers-tasks.js]: replace `bool = false` to ` return false`

File: task/02-numbers-tasks.js [hunk 9]
```diff
@@ -203,7 +198,15 @@ function roundToPowerOfTen(num, pow) {
...
+  }
+
+  return bool;
```
Comment (RamanRudski) [inline task/02-numbers-tasks.js]: replace return like `n !== 1 && n !== 0`

File: task/03-date-tasks.js [hunk 10]
```diff
@@ -75,14 +71,15 @@ function isLeapYear(date) {
  *    Date(2000,1,1,10,0,0),  Date(2000,1,1,15,20,10,453)   => "05:20:10.453"
  */
 function timeSpanToString(startDate, endDate) {
-  throw new Error('Not implemented');
+  const str = new Date(endDate - startDate).toISOString();
+  return str.match(/\d{2}:\d{2}:\d{2}.\d{1,3}/);
```
Comment (RamanRudski) [inline task/03-date-tasks.js]: you can use `str.slice()` it is easier

File: task/04-arrays-tasks.js [hunk 11]
```diff
@@ -37,10 +35,9 @@ function findElement(arr, value) {
  *    5 => [ 1, 3, 5, 7, 9 ]
  */
 function generateOdds(len) {
-  throw new Error('Not implemented');
+  return [...new Array(len)].fill(2).map((e, i) => e * i + 1);
```
Comment (RamanRudski) [inline task/04-arrays-tasks.js:38]: you needn't spread `[...]` just use `new Array(len).fill(.....`

File: task/04-arrays-tasks.js [hunk 12]
```diff
@@ -70,7 +66,7 @@ function doubleArray(arr) {
  *    [] => []
  */
 function getArrayOfPositives(arr) {
-  throw new Error('Not implemented');
+  return arr.filter(e => Math.sign(e) > 0);
```
Comment (RamanRudski) [inline task/04-arrays-tasks.js]: you needn't spread `Math.sign` just use `arr.filter(e => e > 0)`

File: task/04-arrays-tasks.js [hunk 13]
```diff
@@ -102,7 +98,7 @@ function getArrayOfStrings(arr) {
  *    [ false, 0, NaN, '', undefined ]   => [ ]
  */
 function removeFalsyValues(arr) {
-  throw new Error('Not implemented');
+  return arr.filter(e => !!e);
```
Comment (RamanRudski) [inline task/04-arrays-tasks.js]: this solution works without `!!`. remove it

File: task/04-arrays-tasks.js [hunk 14]
```diff
@@ -176,10 +170,9 @@ function getHead(arr, n) {
  *    [ 'a', 'b', 'c', 'd'], 3  => [ 'b', 'c', 'd' ]
  */
 function getTail(arr, n) {
-  throw new Error('Not implemented');
+  return arr.splice(arr.length - n, n);
```
Comment (RamanRudski) [inline task/04-arrays-tasks.js]: you should use `arr.splice(-n)`

File: task/04-arrays-tasks.js [hunk 15]
```diff
@@ -235,7 +227,7 @@ function toArrayOfSquares(arr) {
  *   [ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 ] => [ 1, 3, 6, 10, 15, 21, 28, 36, 45, 55 ]
  */
 function getMovingSum(arr) {
-  throw new Error('Not implemented');
+  return arr.map((e => el => (e += el))(0));
```
Comment (RamanRudski) [inline task/04-arrays-tasks.js]: better use reduce in this case

File: task/04-arrays-tasks.js [hunk 16]
```diff
@@ -439,7 +432,11 @@ function sortCitiesArray(arr) {
  *           [0,0,0,0,1]]
  */
 function getIdentityMatrix(n) {
-  throw new Error('Not implemented');
+  return [...new Array(n)].fill([]).reduce((acc, el, index) => {
+    acc.push([...new Array(n)].fill(0));
```
Comment (RamanRudski) [inline task/04-arrays-tasks.js]: replace `[...new Array(n)]` to `new Array(n)`

File: task/04-arrays-tasks.js [hunk 17]
```diff
@@ -471,7 +468,7 @@ function getIntervalArray(start, end) {
  *   [ 1, 1, 2, 2, 3, 3, 4, 4] => [ 1, 2, 3, 4]
  */
 function distinct(arr) {
-  throw new Error('Not implemented');
+  return arr.filter((el, i, array) => array.indexOf(el) === i);
```
Comment (RamanRudski) [inline task/04-arrays-tasks.js]: you can use it `[...new Set([ 1, 1, 2, 2, 3, 3, 4, 4])]`

File: task/04-arrays-tasks.js [hunk 18]
```diff
@@ -505,10 +502,15 @@ function distinct(arr) {
  *   }
  */
 function group(array, keySelector, valueSelector) {
-  throw new Error('Not implemented');
+  return array.reduce((map, value) => {
+    let cities = map.get(keySelector(value));
+    if (cities) cities.push(valueSelector(value));
```
Comment (RamanRudski) [inline task/04-arrays-tasks.js:508]: you must use curly brackets when working with `if else`

**Review by RamanRudski — CHANGES_REQUESTED** [review]

File: task/04-arrays-tasks.js (hunk 18 above)
Comment (d90375) [reply task/04-arrays-tasks.js:508]: discussed

**Review by RamanRudski — APPROVED** [review]

**Pull request merged**